libtransact/src/contract/context/key_value.rs

```diff
@@ -0,0 +1,352 @@
+            &key_map
+                .values()
+                .map(ToOwned::to_owned)
+                .collect::<Vec<String>>(),
+        )?;
+
+        let mut deleted_keys = Vec::new();
+        let mut new_entry_lists = Vec::new();
+        let mut delete_lists = Vec::new();
+        key_map.iter().for_each(|(nkey, addr)| {
+            // Fetching the StateEntryList at the corresponding address
+            if let Some(list) = state_entry_lists.get(addr) {
+                // The StateEntry objects will be filtered out of the StateEntryList if it has the
+                // normalized key. This normalized key is added to a list of successfully filtered
+                // entries to be returned.
+                if list.contains(nkey.to_string()) {
+                    let filtered = list
+                        .entries()
+                        .to_vec()
+                        .into_iter()
+                        .filter(|e| e.normalized_key() != nkey)
+                        .collect::<Vec<StateEntry>>();
+                    if filtered.is_empty() {
+                        delete_lists.push(addr.to_string());
+                    } else {
+                        new_entry_lists.push((addr.to_string(), filtered));
+                    }
+                    deleted_keys.push(nkey.to_string());
+                }
+            }
+        });
+        // Delete any StateEntryLists that have an empty list of entries
+        self.context.delete_state_entries(delete_lists.as_slice())?;
+        // Setting the newly filtered StateEntryLists into state using the internal context
+        self.context.set_state_entries(
+            new_entry_lists
+                .iter()
+                .map(|(addr, filtered_list)| {
+                    let new_entry_list = StateEntryListBuilder::new()
+                        .with_state_entries(filtered_list.to_vec())
+                        .build()
+                        .map_err(|err| KeyValueContextError::ProtocolBuildError(Box::new(err)))?;
+                    Ok((addr.to_string(), new_entry_list.into_bytes()?))
+                })
+                .collect::<Result<Vec<(String, Vec<u8>)>, KeyValueContextError>>()?,
+        )?;
+
+        Ok(deleted_keys)
+    }
+
+    /// This function is unsupported in Sawtooth Sabre. An error should be returned.
+    fn add_receipt_data(&self, _data: Vec<u8>) -> Result<(), KeyValueContextError> {
+        Err(KeyValueContextError::TransactionContextError(
+            ContextError::TransactionReceiptError("Operation not supported".to_string()),
+        ))
+    }
+
+    /// This function is unsupported in Sawtooth Sabre. An error should be returned.
+    fn add_event(
+        &self,
+        _event_type: String,
+        _attributes: Vec<(String, String)>,
+        _data: Vec<u8>,
+    ) -> Result<(), KeyValueContextError> {
+        Err(KeyValueContextError::TransactionContextError(
```

Comment:
Why is this not supported? Is it a hold-over from the original Sabre version of this PR?  Same with the above call.

Reply:
I misread a message in a previous question on how to handle these methods as returning an InvalidTransaction error from the KeyValueTransactionContext, rather than from Sabre.